lib/configurationSchema.js

```diff
@@ -0,0 +1,438 @@
+              type: 'string',
+            },
+          },
+        },
+      ],
+    },
+    packageDependencies: {
+      default: ({ config }) =>
+        findPackageDependencies(
+          config.workingDirectory,
+          config.get('importDevDependencies')
+        ),
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'array',
+          items: {
+            type: `string`,
+          },
+        },
+      ],
+    },
+    parserPlugins: {
+      default: DEFAULT_PARSER_PLUGINS,
+      type: 'array',
+      items: {
+        anyOf: [
+          {
+            type: 'string',
+          },
+          {
+            type: 'array',
+            minItems: 2,
+            additionalItems: false,
+            items: [
+              {
+                type: 'string',
+              },
+              {
+                type: 'object',
+                additionalProperties: {
+                  type: 'string',
+                },
+              },
+            ],
+          },
+        ],
+      },
+    },
+    sortImports: {
+      default: true,
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'boolean',
+        },
+      ],
+    },
+    stripFileExtensions: {
+      default: ['.js', '.jsx', '.ts', '.tsx'],
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'array',
+          items: {
+            type: 'string',
+          },
+        },
+      ],
+    },
+    tab: {
+      default: '  ',
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'string',
+        },
+      ],
+    },
+    useRelativePaths: {
+      default: true,
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'boolean',
+        },
+      ],
+    },
+  },
+};
+
+function findGlobalsFromEnvironments(environments) {
+  const result = Object.keys(globals.builtin);
+
+  environments.forEach((environment) => {
+    const envGlobals = globals[environment];
+    if (!envGlobals) {
+      return;
+    }
+    result.push(...Object.keys(envGlobals));
+  });
+  return result;
+}
+
+export function getDefaultConfig() {
+  return Object.entries(SCHEMA.properties).reduce((acc, [k, v]) => {
+    if (typeof v !== `object` || v === null) {
+      throw new Error(`Expected schema key '${k}' to be an object`);
+    }
+    if (v.hasOwnProperty('default')) {
+      acc[k] = v.default;
+    }
+    return acc;
+  }, {});
+}
+
+export function validate(data) {
+  const ajv = new Ajv();
+  ajvInstanceof(ajv);
+
+  const validate = ajv.compile(SCHEMA);
+
+  if (!validate(data)) {
+    return {
+      error: true,
+      messages: validate.errors
+        .map((err) => {
+          // report unknown identifiers the same as we have done in the past
+          // so we can showcase that the move to Ajv didn't change previous
+          // behaviors
```

Comment:
I appreciate your attention to detail but I do wonder how important is it to preserve this exact behavior?

Reply:
Not important at all :smile: But it helps reduce the scope of this PR and lowers my anxiety of having accidentally introduced some unwanted side-effect.

I was thinking of improving error messages in an upcoming PR. 

Faulty identifiers / bad configuration would be very nice to have the information of what the offending value was, and how to fix it. Ajv gives me a lot of duplicate error messages, as well as false positives when using `anyOf`, so I think that stuff will need a bit of care anyway.

Reply:
Improving error messages sounds great to me. I wonder if anything could be added to the schema that would help with that, or if ajv offers any options to tweak the output it generates out of the box. Or maybe someone else has put out a package that makes ajv errors really nice?

Reply:
Good ideas. There must be. I will look into it.

I was also thinking we might move documentation into the JSON schema and auto-generate it. I came across [jsonschema2md](https://github.com/adobe/jsonschema2md) that might be interesting for this. But that's yet another PR.

Reply:
Started #615 for invalid configuration error messages and #616 for auto-generating documentation